File: Problems/0400_0499/0449_Serialize_and_Deserialize_BST/Project_Go/solution/codec2.go
package solution

import (
	"math"
)

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
/*
 The tree can be built using preorder or postorder traversal and inorder traversal.
 since the tree is BST, we already know that inorder traversal can give sorted order.
serialize tree using preorder or postorder traversal
deserialize using preoder/postorder and inorder traversal
*/

// 148ms

type Codec2 struct {

}

func Constructor() Codec2 {
	return Codec2{}
}

// Serializes a tree to a single string.
func (this *Codec2) serialize(root *TreeNode) string {
	if root == nil {
		return ""
	}

	nodes := []rune{}
	var preOrder func(*TreeNode)
	preOrder = func(curr *TreeNode) {
		if curr == nil {
			return
		}
		nodes = append(nodes, rune(curr.Val))
		preOrder(curr.Left)
		preOrder(curr.Right)
	}

	preOrder(root)
	return string(nodes)
}


// Deserializes your encoded data to tree.
func (this *Codec2) deserialize(data string) *TreeNode {
	if len(data) == 0 {
		return nil
	}
	nodes := []rune{}
	for _, n := range data {
		nodes = append(nodes, n)
	}
	var buildTree func([]rune, int, int) *TreeNode
	startIndex := 0
	buildTree = func(nodes []rune, min, max int) *TreeNode {
		if startIndex == len(nodes) {
			return nil
		}
		val := int(nodes[startIndex])
		if val < min || val > max{
			return nil
		}
		root := &TreeNode{val, nil, nil}
		startIndex++
		root.Left = buildTree(nodes, min, val)
		root.Right = buildTree(nodes,val, max)
		return root
	}
	return buildTree(nodes, math.MinInt32, math.MaxInt32)
}


/**
 * Your Codec object will be instantiated and called as such:
 * obj := Constructor();
 * data := obj.serialize(root);
 * ans := obj.deserialize(data);
 */
 

File: Problems/0400_0499/0449_Serialize_and_Deserialize_BST/Project_Java/Codec2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
import java.util.*;

// 7ms

public class Codec2 {
    private static final String SEP = ",";
    private static final String NULL = "null";

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        StringBuilder sb = new StringBuilder();
        if (root == null) {
            return NULL;
        }
        Stack<TreeNode> st = new Stack<>();
        st.push(root);

        while (!st.empty()) {
            root = st.pop();
            sb.append(root.val).append(SEP);
            if (root.right != null) {
                st.push(root.right);
            }
            if (root.left != null) {
                st.push(root.left);
            }
        }

        return sb.toString();
    }

    // Decodes your encoded data to tree.
    // pre-order traversal
    public TreeNode deserialize(String data) {
        if (data.equals(NULL)) {
            return null;
        }

        String[] strs = data.split(SEP);
        Queue<Integer> q = new LinkedList<>();

        for (String e : strs) {
            q.offer(Integer.parseInt(e));
        }

        return getNode(q);
    }

    private TreeNode getNode(Queue<Integer> q) {
        if (q.isEmpty()) {
            return null;
        }

        TreeNode root = new TreeNode(q.poll());
        Queue<Integer> samllerQueue = new LinkedList<>();

        while (!q.isEmpty() && q.peek() < root.val) {
            samllerQueue.offer(q.poll());
        }

        root.left = getNode(samllerQueue);
        root.right = getNode(q);
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// String tree = ser.serialize(root);
// TreeNode ans = deser.deserialize(tree);
// return ans;
